### ISSUE TITLE
Prisma deploy doesn't generate files on windows

### ISSUE THREAD

**Issue by blocka** [prismagraphql/prisma#2494]
# Bug Report

**Current behavior**
`prisma deploy` is not generating files

**Reproduction**
1) install typescript-basic boilerplate
2) delete generated files
3) `prisma deploy`

**Expected behavior?**
Should see generated files again

**Comment by marktani** [comment]
Thanks for the bug report!

It looks like the problem is that `prisma-binding` has been upgraded to `2.x`, but `graphql prepare` is still used instead of `graphql codegen`.

We'll update this soon!

**Comment by marktani** [comment]
The problem is that `.graphqlconfig.yml` is not correctly configured :slightly_smiling_face:

**Comment by yocheved** [comment]
```
projects:
  app:
    schemaPath: src/schema.graphql
    extensions:
      endpoints:
        default: http://localhost:4000
  database:
    schemaPath: src/generated/prisma.graphql
    extensions:
      prisma: database/prisma.yml
      prepare-binding:
        output: src/generated/prisma.ts
        generator: prisma-ts
```

What is incorrect in there? Is there something I can change that will make this work?

**Comment by yocheved** [comment]
To clarify, when I run `prisma deploy`  it does show this outcome 
```
post-deploy:
Running graphql get-schema --project database √
Running graphql prepare √
``` 
but no files are generated.
If I run those commands each on their own in the cmd prompt, that does generate the files.

**Comment by nikolasburk** [comment]
Sorry for the confusion, the code generation for prisma-binding 2.0 works slightly differently as it's not using the `graphql prepare` but the new `graphql codegen` command instead. 

You basically need to update your `prisma.yml` and the `.graphqlconfig.yml`:

 **prisma.yml**

```yml
# ... other properties

hooks:
  post-deploy:
    - graphql get-schema
    - graphql codegen
```

**.graphqlconfig.yml**

```yml
projects:
  myapp:
    schemaPath: src/generated/prisma.graphql
    extensions:
      prisma: prisma/prisma.yml
      codegen:
        - generator: prisma-binding
          language: typescript
          output:
            binding: src/generated/prisma.ts
```

Let me know if you have any further questions 🙌

**Comment by nikolasburk** [comment]
Hmm, I see. In that case I'm reopening the issue for someone to look into that specific case. Thanks @yocheved 🙏

**Comment by marktani** [comment]
@yocheved can you provide reproduction repositories for both cases? :slightly_smiling_face:

**Comment by blocka** [comment]
I was able to reproduce this directly from the typescript-basic boilerplate.

**Comment by blocka** [comment]
Sorry for the delay,
It is still an issue.

**Comment by gsancho** [comment]
i have the same problem :(

**Comment by fonzied** [comment]
I believe prisma deploy is not running `graphql get-schema --project database` properly. Or maybe it's not passing the users .env flags. But regardless, if I run `prisma deploy` then manually run `graphql get-schema --project database --dotenv=myenvpath`, it generates the file properly. So is this just a script error?

**Comment by ntziolis** [comment]
@marktani This problem still persists after latest upgrades to boilerplate projects:
- tried again yesterday evening
- schema gets downloaded properly
- bindings are reportedly created but that's not true, no file is written anywhere during bindings creation

Getting the following warnings
`warning command prepare both exists in plugin C:\Users\nikol\Documents\git\circula\server\node_modules\graphql-cli-prepare and is shipped with the graphql-cli.`

Basically that leaves us not being able to migrate to 1.7+ right now.

**Comment by marktani** [comment]
Thanks for your feedback @ntziolis! Can you share your `.graphqlconfig.yml` and the hooks from `prisma.yml`?

**Comment by ntziolis** [comment]
@marktani Was in the rpocess of updating my post

**Comment by marktani** [comment]
Does it work if you run `graphql codegen` directly from the command line, and not as a hook?

**Comment by ntziolis** [comment]
Nope, same result reports success + warning. Even tried with specific project but same result:
![image](https://user-images.githubusercontent.com/265338/40533612-977d2410-6003-11e8-84b7-b70afb93b62b.png)

**Comment by ntziolis** [comment]
Basically im out of ideas. The problematic package in question is actually a dep of graphql-cli, so removing this would have to be done manually after each yarn install which is NOT a viable option:
![image](https://user-images.githubusercontent.com/265338/40535818-7ce7421e-600a-11e8-8111-f799f2d094c8.png)

if anybody has any ideas how to resolve this issue please let me as we are pretty much blocked from moving forward.

**Comment by ntziolis** [comment]
I have taken some time to summarize this issue as I think this are multiple:
1. (solved) New way the code generation gets triggered <= solved
2. (solved) Codegen has to be used instead of prepare, which was wrong in the old boilerplates <= solved
3. (open) `prisma deploy` does not properly trigger running scripts (at least on windows) <= OPEN
4. (open) While `graphql get-schema' works when manually executed `graphql codegen `command does NOT work on windows when manually executed

Summary
I think we can live with no 3 for a bit, but NOT with No 4 as this basically requires windows folks to stay below version 1.7.

**Comment by marktani** [comment]
Thanks for the summary, @ntziolis. We will work on this issue soon! :slightly_smiling_face:

---

One note, `graphql codegen` is tied to `prisma-binding` version `2.x`, while `graphql prepare` is tied to `prisma-binding` version `1.5.x`.

Neither is tied to the `prisma` version itself, so if you want to upgrade to `prisma` 1.7+, you can do so on Windows as well, if you use `prisma-binding` version `1.5.x` and `graphql prepare` :slightly_smiling_face:

**Comment by ntziolis** [comment]
Thank you for the workaround. We usually try and avoid using "unusual" by staying with the documented setup paths to not not let debt aggregate over time, but in this case I don't see another way. Thx again.

**Comment by divyenduz** [comment]
I am able to replicate this issue (`prisma-binding` 2.0 with `graphql codegen`) on windows, and on mac machines when I remove the `node_modules` folder. 

I have created an [issue](https://github.com/graphql-cli/graphql-cli/issues/306) with related findings.

**Comment by zjyonline** [comment]
Prisma deploy doesn't generate files on mac.
My graphqlconfig.yml content is 
  prisma:
    schemaPath: src/generated/prisma.graphql
    extensions:
      prisma: prisma/prisma.yml
prisma.yml is 
hooks:
  post-deploy:
    - graphql get-schema --project prisma
and I run deploy command, there is no file named 'prisma.graphql' created.

**Comment by jhalborg** [comment]
@marktani - Can you explain that workaround from [your previous comment](https://www.av-cables.dk/smartphone-bilholder/universal-bilholder-med-sugekop.html) for people who've just started with the advanced boilerplate and `prisma 1.8` and `prisma-binding 2.x.x` ? What should the post-deploy hook look like, and what packages and versions should be used while we wait for a fix here?

**Comment by jhalborg** [comment]
@williamluke4 - Sorry, my bad 😆  Fixed

**Comment by zjyonline** [comment]
I solve my problem with command
`npm install -g graphql-cli`
then I deploy my project,show that: 
`project prisma - Schema file was created: src/generated/prisma.graphql`

**Comment by yocheved** [comment]
@zjyonline that didn't fix it for me

**Comment by ntziolis** [comment]
@zjyonline didn't fix it for me either.

**Comment by zjyonline** [comment]
I found that hooks command run in this file in my Mac ,
`/usr/local/lib/node_modules/prisma/node_modules/prisma-cli-core/dist/commands/deploy/index.js`
about lin4 400,
        ```
case 11:
                        hooks = this.definition.getHooks('post-deploy');
                        if (hooks.length > 0) {
                            this.out.log("\n" + chalk_1.default.bold('post-deploy') + ":");
                        }
                        for (_i = 0, hooks_1 = hooks; _i < hooks_1.length; _i++) {
                            hook = hooks_1[_i];
                            splittedHook = hook.split(' ');
                            this.out.action.start("Running " + chalk_1.default.cyan(hook));
                            child = **npm_run_1.spawnSync(splittedHook[0], splittedHook.slice(1));**
                            stderr = child.stderr && child.stderr.toString();
                            if (stderr && stderr.length > 0) {
                                this.out.log(stderr);
                            }
                            stdout = child.stdout && child.stdout.toString();
                            if (stdout) {
                                this.out.log(stdout);
                            }
                            this.out.action.stop();
                        }
                        return [2 /*return*/];
```
So I found maybe I forget to install something about 'graphql-cli' ?
@yocheved @ntziolis read this file maybe help.

**Comment by jhalborg** [comment]
Any ETA on this? It's blocking half our development team 😄

**Comment by divyenduz** [comment]
@ntziolis : Hello, you are kinda correct, we are [working on a solution](https://github.com/graphql-cli/graphql-cli/pull/311) and we should be able to release it soon.

**Comment by divyenduz** [comment]
This is released in [`graphql-cli` version 2.6.1](https://github.com/graphql-cli/graphql-cli/releases/tag/v2.16.1).

Please update the CLI and try it out. Then we can close this issue.

**Comment by ntziolis** [comment]
Wow 20min to close it. Will test as soon as in front of pc!
		
		

		
	




On Fri, Jun 1, 2018 at 10:57 AM +0200, "Divyendu Singh" <notifications@github.com> wrote:












This is released in graphql-cli version 2.6.1.


Please update the CLI and try it out. Then we can close this issue.

**Comment by williamluke4** [comment]
Just tested with ```"prisma-binding": "^2.0.2"``` and ``` "graphql-cli": "^2.16.1"``` and is working! Thanks guys

**Comment by ntziolis** [comment]
I can confirm that `graphql codegen` now works with `graphql-cli@2.6.1` + `prisma-bindings@2.0.2`.

Updating the summary I made previously:
- (closed) `graphql codegen` command does NOT work on windows
- (still open) `prisma deploy` does not actually execute the post deployment scripts (at least on windows) while it claims the scripts were successfully executed

Assuming the error was in the way `graphql codegen` spawns `apollo-codegen`. It should basically be the same fix for prisma deploy which spawns `graphql get-schema` + `graphql codegen`.

**Comment by divyenduz** [comment]
@ntziolis : Can you please share your graphql config and `prisma.yml` file and prisma cli version via `prisma --version`?

**Comment by ntziolis** [comment]
@divyenduz I understand the changes and have implemented them @marktani has already double checked them above.

To be clear:
- The scripts supposedly get executed (they are listed as successfully executed in the console including a checkmark)
- but the scripts never **actually** get executed
- to check simply delete the resulting files and rerun .. the files will not be generated


Here is what my console looks like:
```
# prisma deploy
Deploying service `circula` to stage `dev` to server `local` 167ms
Service is already up to date.

post-deploy:
Running graphql get-schema --project database √
Running graphql codegen --project database √
```

=> No files get created, neither prisma.ts nor schema.graphql

**Comment by divyenduz** [comment]
🎉This is released in prisma cli [version 1.9](https://github.com/prismagraphql/prisma/releases/tag/1.9). 🎉

Please download using `npm install -g prisma`

**Comment by ntziolis** [comment]
@divyenduz Right now this does not work, when installing prisma it only installs 1.8.3. npm side says this is the most recent version. did something go wrong during publishing to npm?

**Comment by marktani** [comment]
Yes, we're looking into publishing 1.9 to npm tomorrow.

**Comment by abelovic** [comment]
When you create a release it should be published to npm at the same time as part of your deployment process always

**Comment by marktani** [comment]
Yes, we're setting up an automatic publishing flow as you suggested :slightly_smiling_face:

**Comment by rohmanhm** [comment]
I still get the error message, but everything worked well.

I'm on Windows 10

Here the error message
```bash
Romsy@ROMSY-PC MINGW64
$ prisma deploy
Deploying service `default` to stage `default` to server `default` 48ms
Service is already up to date.

post-deploy:
warning command prepare both exists in plugin pathsomewhere\node_modules\graphql-cli-prepare and is shipped with the graphql-cli.
The plugin is being ignored.

Running graphql get-schema --project database...
project database - No changes
Running graphql get-schema --project database √
warning command prepare both exists in plugin pathsomewhere\node_modules\graphql-cli-prepare and is shipped with the graphql-cli.
The plugin is being ignored.

Running graphql codegen √
```

Here's my dep and devDep
```js
  "dependencies": {
    "bcryptjs": "2.4.3",
    "graphql-yoga": "1.14.10",
    "jsonwebtoken": "8.3.0",
    "prisma-binding": "2.1.0"
  },
  "devDependencies": {
    "@types/bcryptjs": "2.4.1",
    "dotenv-cli": "1.4.0",
    "graphql-cli": "2.16.3",
    "nodemon": "1.17.5",
    "npm-run-all": "4.1.3",
    "prisma": "1.10.2",
    "rimraf": "2.6.2",
    "ts-node": "6.2.0",
    "typescript": "2.9.2"
  }
```

**Comment by yocheved** [comment]
@rohmanhm I also get that warning but it everything still works as expected

**Comment by konsumer** [comment]
I am still getting this with prisma 1.15.3 & graphql-cli 2.16.5.

If I run `PRISMA_ENDPOINT=http://localhost:4466 prisma deploy`:

```
Deploying service `default` to stage `default` to server `local` 45ms
Service is already up to date.

post-deploy:
Running graphql get-schema --project database ✔
```

but `server/api/database/generated.graphql` isn't created.

If I run `PRISMA_ENDPOINT=http://localhost:4466 graphql get-schema --project database`

it generates it, fine.

`.graphqlconfig.yml`:
```
projects:
  app:
    schemaPath: server/api/schema/*.graphql
    extensions:
      endpoints:
        default: http://localhost:8000
  
  database:
    schemaPath: server/api/database/generated.graphql
    extensions:
      prisma: server/api/database/prisma.yml
```

`server/api/database/prisma.yml`
```
endpoint: ${env:PRISMA_ENDPOINT}

datamodel: datamodel.graphql

seed:
  import: seed.graphql

hooks:
  post-deploy:
    - graphql get-schema --project database
```

**Comment by juancamiloq1** [comment]
> I managed to get it to work right by installing graphql "^0.13.0", instead of "^0.14.0".

Same for me. It works when I change the grapqhl version

**Comment by JensMadsen** [comment]
cannot get it to work even when downgrading graphql version as suggested

**Comment by pitops** [comment]
i can confirm version 0.14.0 is not working to generate schema

**Comment by steph4nc** [comment]
I also have the same problem, running on "graphql": "^0.13.2"

**Comment by pitops** [comment]
I find a working solution so far

add this to your prisma.yml file

```yaml
hooks:
  post-deploy:
    - graphql get-schema -p prisma
    - prisma generate
generate:
  - generator: graphql-schema
    output: ./src/generated/
```

**Comment by saifali96** [comment]
I'm following the GraphQL tutorial from [How To GraphQL](https://www.howtographql.com/graphql-js/5-connecting-server-and-database/) and when I run `prisma deploy` on Windows, it doesn't generate the schema.

Output as follow:

> Deploying service `XXX-node` to stage `dev` to server `prisma-eu1` 186ms
> Service is already up to date.
> post-deploy:
> spawnSync graphql ENOENT
> Running graphql get-schema --project db ×

I tried to change the post deploy code in many ways, nothing seemed to work. Even simple echo messages are not being executed also I added a silent command to create a file but event that wasn't executed. 

My observation is that this is a scripting issue and the arguments are never passed to `powershell `or `cmd `for execution.

Workaround is to run the `graphql get-schema --project db` yourself which does the job.

My environment and version is latest, created and fetched/installed today.

**Comment by saifali96** [comment]
@rohmanhm nice idea, I changed my laptop and almost forgot about mingw. Thanks I'll use that.
But given the number of comments on this issue, I think this should be addressed in the next release.

Cheers!

**Comment by SeanK1191** [comment]
Did using git bash work for you @saifali96? I'm having the same problem, thought I had it working but now I'm just getting same output as you had in cmd and git bash.

**Comment by SeanK1191** [comment]
Also getting same problem as @saifali96. 

Did Git Bash fix it for you? It's giving me the same error and nothing generated.

**Comment by saifali96** [comment]
@SeanK1191 no. Unfortunately it *doesn't* work with `git bash` either.

For now, only workaround is to run the command yourself in the terminal.

**Comment by SeanK1191** [comment]
@divyenduz I believe issue #3491 is the same or related to the issue mentioned here.

@saifali96 I'm also doing the HowToGraphQL tutorial, gonna remove the post hook and use this npm script for now, hope it helps: 
```
"prisma-deploy": "prisma deploy & graphql get-schema --project database"
```

**Comment by divyenduz** [comment]
Thanks all for reporting this, in the mean time we will downgrade `npm-run` shortly:- https://github.com/prisma/prisma/pull/3517

**Comment by alfredringstad** [comment]
Until this is released, if you use `yarn` a workaround is to add this to your package.json:
```
"resolutions": {
  "npm-run": "4.1.2"
}
```

**Comment by divyenduz** [comment]
This is already available in the latest beta, please install it via `npm install -g prisma@beta`

**Comment by Daniely2564** [comment]
> I believe prisma deploy is not running `graphql get-schema --project database` properly. Or maybe it's not passing the users .env flags. But regardless, if I run `prisma deploy` then manually run `graphql get-schema --project database --dotenv=myenvpath`, it generates the file properly. So is this just a script error?

you are god

**Comment by superhawk610** [comment]
Changing `graphql get-schema --project database` to just `graphql get-schema` fixed it for me, though I'm very new to graph-ql still so it could just be due to an improperly configured project.

**Comment by ShadyMohammed** [comment]
I have the same issue. I'm working on ubuntu 18.10

The output is:
post-deploy:
Running graphql get-schema ✔
Running graphql codegen ✔

but the generated file is never updated.

I have followed all the steps in this thread.

**Comment by smithmanny** [comment]
If anybody is still having trouble adding `.graphql` to the end of your output worked for me.

```
endpoint: ${env:PRISMA_ENDPOINT}
datamodel: datamodel.prisma
# secret: ${env:PRISMA_SECRET}

generate:
  - generator: graphql-schema
    output: src/generated/prisma-client.graphql

hooks:
  post-deploy:
    - prisma generate
```
